File: samples/java/src/main/java/com/example/BankAccountSurgeModel.java
// Copyright © 2017-2021 UKG Inc. <https://www.ukg.com>

package com.example;

import com.example.account.BankAccount;

import com.example.event.BankAccountEvent;
import com.example.format.SurgeAggregateReadFormattingBankAccount;
import com.example.format.SurgeAggregateWriteFormattingBankAccount;
import com.example.format.SurgeEventWriteFormattingBankEvent;
import surge.core.SurgeAggregateReadFormatting;
import surge.core.SurgeAggregateWriteFormatting;
import surge.core.SurgeEventWriteFormatting;
import surge.javadsl.command.AggregateCommandModel;
import surge.javadsl.command.SurgeCommandBusinessLogic;
import surge.kafka.KafkaTopic;
import java.util.UUID;

public class BankAccountSurgeModel extends SurgeCommandBusinessLogic<UUID, BankAccount, BankAccountCommand,
        BankAccountEvent> {
    @Override
    public AggregateCommandModel<BankAccount, BankAccountCommand, BankAccountEvent> commandModel() {
        return new BankAccountCommandModel();
    }

    @Override
    public String aggregateName() {
        return "bank-account";
    }

    @Override
    public KafkaTopic stateTopic() {
        return new KafkaTopic("bank-account-state");
    }

    @Override
    public KafkaTopic eventsTopic() {
        return new KafkaTopic("bank-account-events");
    }

    @Override
    public SurgeAggregateReadFormatting<BankAccount> aggregateReadFormatting() {
        return new SurgeAggregateReadFormattingBankAccount();
    }

    @Override
    public SurgeEventWriteFormatting<BankAccountEvent> eventWriteFormatting() {
        return new SurgeEventWriteFormattingBankEvent();
    }

    @Override
    public SurgeAggregateWriteFormatting<BankAccount> aggregateWriteFormatting() {
        return new SurgeAggregateWriteFormattingBankAccount();
    }
}

File: samples/scala/src/main/scala/com/example/model/BankAccountSurgeModel.scala
package com.example.model

import play.api.libs.json.Json
import surge.core.{SerializedAggregate, SerializedMessage, SurgeAggregateReadFormatting, SurgeAggregateWriteFormatting, SurgeEventWriteFormatting}
import surge.kafka.KafkaTopic
import surge.scaladsl.command.{AggregateCommandModel, SurgeCommandBusinessLogic}
import com.example.account.BankAccount
import com.example.command._
import com.example.event._

import java.util.UUID

object BankAccountSurgeModel extends SurgeCommandBusinessLogic[UUID, BankAccount, BankAccountCommand, BankAccountEvent] {
  override def commandModel: AggregateCommandModel[BankAccount, BankAccountCommand, BankAccountEvent] = BankAccountCommandModel

  override def aggregateName: String = "bank-account"

  override def stateTopic: KafkaTopic = KafkaTopic("bank-account-state")

  override def eventsTopic: KafkaTopic = KafkaTopic("bank-account-events")

  override def aggregateReadFormatting: SurgeAggregateReadFormatting[BankAccount] = (bytes: Array[Byte]) => Json.parse(bytes).asOpt[BankAccount]

  override def aggregateWriteFormatting: SurgeAggregateWriteFormatting[BankAccount] = (agg: BankAccount) => {
    val aggBytes = Json.toJson(agg).toString().getBytes()
    val messageHeaders = Map("aggregate_id" -> agg.accountNumber.toString)
    SerializedAggregate(aggBytes, messageHeaders)
  }

  override def eventWriteFormatting: SurgeEventWriteFormatting[BankAccountEvent] = (evt: BankAccountEvent) => {
    val evtKey = evt.accountNumber.toString
    val evtBytes = evt.toJson.toString().getBytes()
    val messageHeaders = Map("aggregate_id" -> evt.accountNumber.toString)
    SerializedMessage(evtKey, evtBytes, messageHeaders)
  }
}